streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java

```diff
@@ -52,44 +52,41 @@
     private final Serde<K> keySerde;
 
     private final Serde<V> valueSerde;
-    private String name;
-    protected Map<K, V> map;
+    private final String name;
+    protected final Map<K, V> map;
+
     private StateSerdes<K, V> serdes;
+    private boolean restoring = false;      // TODO: this is a sub-optimal solution to avoid logging during restoration.
+                                            // in the future we should augment the StateRestoreCallback with onComplete etc to better resolve this.
     private volatile boolean open = true;
 
-    protected EldestEntryRemovalListener<K, V> listener;
+    EldestEntryRemovalListener<K, V> listener;
 
-    // this is used for extended MemoryNavigableLRUCache only
-    public MemoryLRUCache(Serde<K> keySerde, Serde<V> valueSerde) {
+    MemoryLRUCache(String name, final int maxCacheSize, Serde<K> keySerde, Serde<V> valueSerde) {
+        this.name = name;
         this.keySerde = keySerde;
         this.valueSerde = valueSerde;
-    }
-
-    public MemoryLRUCache(String name, final int maxCacheSize, Serde<K> keySerde, Serde<V> valueSerde) {
-        this(keySerde, valueSerde);
-        this.name = name;
 
         // leave room for one extra entry to handle adding an entry before the oldest can be removed
         this.map = new LinkedHashMap<K, V>(maxCacheSize + 1, 1.01f, true) {
             private static final long serialVersionUID = 1L;
 
             @Override
             protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
-                if (super.size() > maxCacheSize) {
-                    K key = eldest.getKey();
-                    if (listener != null) listener.apply(key, eldest.getValue());
-                    return true;
+                boolean evict = super.size() > maxCacheSize;
+                if (evict && !restoring && listener != null) {
```

Comment:
do the updates on `restoring` from the `StateRestoreCallback` always happen in the same thread?